File: api-common/src/main/java/com/ecsimsw/common/support/aop/MemLock.java
package com.ecsimsw.common.support.aop;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;

public class MemLock {

    private static final ConcurrentHashMap<String, AtomicBoolean> LOCKS = new ConcurrentHashMap<>();

    public static void tryLock(String key, long timeoutMs) throws InterruptedException {
        var locked = LOCKS.computeIfAbsent(key, k -> new AtomicBoolean(false));
        var start = System.currentTimeMillis();
        while (System.currentTimeMillis() - start < timeoutMs) {
            if (locked.compareAndSet(false, true)) {
                return;
            }
            Thread.sleep(10);
        }
        throw new IllegalStateException("Failed to acquire lock");
    }

    public static void unlock(String key) {
        LOCKS.remove(key);
    }
}

File: _api-transaction/src/main/kotlin/com/ecsimsw/transaction/support/MemLock.kt
package com.ecsimsw.transaction.support

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean

object MemLock {

    private val LOCKS = ConcurrentHashMap<String, AtomicBoolean>()

    fun tryLock(key: String, timeoutMs: Long) {
        val locked = LOCKS.computeIfAbsent(key) { AtomicBoolean(false) }
        val start = System.currentTimeMillis()
        while (System.currentTimeMillis() - start < timeoutMs) {
            if (locked.compareAndSet(false, true)) {
                return
            }
            Thread.sleep(10)
        }
        throw IllegalStateException("Failed to acquire lock")
    }

    fun unlock(key: String) {
        LOCKS.remove(key)
    }
}